File: Java/Language/Basics/InterfaceTest2.java
interface Consumer {

	void apply(int action);

	default void applyAll(int count) {
		for(int a = 1; a <= count; ++a)
			apply(a);
	}
}

class SomeResource implements Consumer, AutoCloseable {

	private String name;

	static { //static initializer - is invoked once automatically when the class is used for the first time in a program
		System.out.println("SomeResource class initialized.");
	}

	public SomeResource(String id) {
		name = id;
		System.out.printf("Acquired %s resource%n", name);
	}

	public void apply(int action) {
		System.out.printf("Action %d performed on %s resource%n", action, name);
	}

	public void close() {
		System.out.printf("Releasing %s resource%n", name);
	}
}


class InterfaceTest2 {

	private static void run(String info) {
		/*
		var c = new SomeResource("third");
		try{
			c.applyAll(Integer.parseInt(info));
		}finally{
			c.close();
		}
		*/
		try(var c = new SomeResource("third")){
			c.applyAll(Integer.parseInt(info));
		}
	}

	public static void main(String[] args) {
		var a = new SomeResource("first");
		a.apply(1);
		a.close();
		var b = new SomeResource("second");
		b.apply(2);
		b.close();
		try{
			run(args[0]);
		}catch(Exception e){}
	}
}

File: CSharp/Basics/InterfaceTest2.cs
using System;

interface IConsumer : IDisposable
{
	void Apply(int action);

	void ApplyAll(int count)
	{
		for(int a = 1; a <= count; ++a)
			Apply(a);
	}
}

class ResourceConsumer : IConsumer
{
	private string id;

	static ResourceConsumer()
	{
		Console.WriteLine("ResourceConsumer type initialized");
	}

	internal ResourceConsumer(string name)
	{
		id = name;
		Console.WriteLine($"{id} resource acquired");
	}

	void IConsumer.Apply(int action)
	{
		if(id != null)
			Console.WriteLine($"{id} resource consumed with action {action}");
	}

	void IDisposable.Dispose()
	{
		if(id != null)
		{
			Console.WriteLine($"{id} resource released");
			id = null;
		}
	}
}	

static class Program
{
	private static void Run(string cmd)
	{
		using(IConsumer c = new ResourceConsumer("third"))
		{
			c.ApplyAll(int.Parse(cmd));
		}
	}

	public static void Main(string[] args)
	{
		IConsumer a = new ResourceConsumer("first");
		a.Apply(1);
		a.Dispose();
		IConsumer b = new ResourceConsumer("second");
		b.Apply(2);
		b.Dispose();
		try
		{
			Run(args[0]);
		}
		catch {}
	}
}

